addons/mail/static/src/owl/hooks/use_store_compare_keys.js

```diff
@@ -0,0 +1,121 @@
+     * calling `isEqual` on primitive values and objects, but which also works for
+     * Owl Proxy in a temporal way: the current result of hashFn is compared to the
+     * previous result of hashFn (from the last time the function was called).
+     *
+     * It means that when this function is given Proxy the first call will always
+     * return false, and consecutive calls will not lead to the same result:
+     * it returns true the first time after a change happended inside the Proxy, and
+     * then always returns false until a new change is made.
+     *
+     * @returns {function} function taking two arguments, and comparing them as
+     *  explained above.
+     */
+    function proxyComparator() {
+        /**
+         * It is important to locally save the old `revNumber` of each resulting
+         * value because when the "old" and "new" values are the same proxy it is
+         * impossible to compare them based on their current value (since it was
+         * updated in "both" due to the fact it is a proxy in the first place).
+         *
+         * And if the values are not proxy, `revNumber` will be 0 and the `isEqual`
+         * will be used to compare them.
+         */
+        let oldRevNumber;
+
+        function compare(a, b) {
+            let ok = true;
+            const newRevNumber = hashFn(b);
+            if (a === b && newRevNumber > 0) {
+                ok = oldRevNumber === newRevNumber;
+            } else {
+                ok = isEqual(a, b);
+            }
+            oldRevNumber = newRevNumber;
+            return ok;
+        }
+
+        return compare;
+    }
+
+    /**
+     * @see proxyComparator, but instead of comparing the given values, it compares
+     * their respective keys, with `compareDepth` level of depth.
+     * 0 = compare key, 1 = also compare subkeys, ...
+     *
+     * This assumes the given values are objects or arrays.
+     *
+     * @param {number|object} compareDepth the comparison depth, either as number
+     *  (applies to all keys) or as an object (depth for specific keys)
+     * @returns {function}
+     */
+    function proxyComparatorDeep(compareDepth = 0) {
+        const comparator = proxyComparator();
+        const comparators = {};
+
+        function compare(a, b) {
+            // If a and b are (the same) proxy, it is already managing the depth
+            // by itself, and a simple comparator can be used.
+            if (a === b && hashFn(b) > 0) {
+                return comparator(a, b);
+            }
+            let ok = true;
+            const newKeys = Object.keys(b);
+            if (Object.keys(a).length !== newKeys.length) {
+                ok = false;
+            }
+            for (const key of newKeys) {
+                // the depth can be given either as a number (for all keys) or as
+                // an object (for each key)
+                const depth = typeof compareDepth === 'number' ? compareDepth : compareDepth[key] || 0;
```

Comment:
line too long, would put brackets because not sure of precedence of `||` and `? :`

Reply:
ok, I will use if/else